codec-dns-doh/src/main/java/io/netty/handler/codec/dns/doh/DoHQueryEncoder.java

```diff
@@ -0,0 +1,121 @@
+
+    private final DnsQueryEncoder encoder;
+    private final boolean http2;
+    private final boolean get;
+    private final URL url;
+
+    /**
+     * Creates a new encoder with {@linkplain DnsRecordEncoder#DEFAULT the default record encoder},
+     * uses HTTP/1.1 and HTTP POST method.
+     *
+     * @param url DoH Upstream Server
+     */
+    public DoHQueryEncoder(URL url) {
+        this(DnsRecordEncoder.DEFAULT, false, false, url);
+    }
+
+    /**
+     * Creates a new encoder with {@linkplain DnsRecordEncoder#DEFAULT the default record encoder}, uses HTTP POST
+     * method and specifies if we're using HTTP/2 (h2).
+     *
+     * @param http2 Use HTTP/2 (h2)
+     * @param url   DoH Upstream Server
+     */
+    public DoHQueryEncoder(boolean http2, URL url) {
+        this(DnsRecordEncoder.DEFAULT, http2, false, url);
+    }
+
+    /**
+     * Creates a new encoder with the specified {@code recordEncoder}, {@code HTTP2},
+     * {@code GET} and {@code url}
+     *
+     * @param recordEncoder DNS Record Encoder
+     * @param http2         Use HTTP/2 (h2)
+     * @param get           Use HTTP GET method
+     * @param url           DoH Upstream Server
+     */
+    public DoHQueryEncoder(DnsRecordEncoder recordEncoder, boolean http2, boolean get, URL url) {
+        this.encoder = new DnsQueryEncoder(recordEncoder);
+        this.http2 = http2;
+        this.get = get;
+        this.url = url;
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, DefaultDnsQuery msg, List<Object> out) throws Exception {
+        ByteBuf byteBuf = ctx.alloc().buffer();
+        encoder.encode(msg, byteBuf);
+
+        FullHttpRequest fullHttpRequest;
+        if (get) {
+            ByteBuf encodedDNS = Base64.encode(byteBuf, Base64Dialect.URL_SAFE);
+
+            /*
+             * As per RFC 8484, variable "dns" is specified for GET request.
+             * [https://tools.ietf.org/html/rfc8484#section-4.1]
+             */
+            fullHttpRequest = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET,
+                    url.getPath() + "?dns=" + encodedDNS.toString(CharsetUtil.UTF_8));
+
+            // Release ByteBuf since we only use `toString()` method to prevent memory leak.
+            encodedDNS.release();
+        } else {
+            fullHttpRequest = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST,
+                    url.getPath(), byteBuf);
+            fullHttpRequest.headers().add(HttpHeaderNames.CONTENT_LENGTH, byteBuf.readableBytes());
+        }
+
+        fullHttpRequest.headers()
+                .add(HttpHeaderNames.HOST, url.getHost())
+                .add(HttpHeaderNames.CONTENT_TYPE, "application/dns-message")
+                .add(HttpHeaderNames.ACCEPT, "application/dns-message");
+
+        // If we're using HTTP/2 (h2) then we'll add "x-http2-scheme" header
+        if (http2) {
```

Comment:
this depends on the fact that would have the adapter in the pipeline. I wonder if we should just not support http2 for now

Reply:
or maybe we should better use `Http2Frame`s

Reply:
The thing is, HTTP/2 is RECOMMENDED by RFC for DoH. I am looking for a better approach to handle this.

Reply:
+1 to HTTP/2, I think you need 2 http requests to serve a normal DNS query.  And there is hopefully a high chance of reuse depending on number of distinct host lookups and the TTL.